global: Allow updates of failed services with restart policy "none"

aaronlehmann opened the pull request:
There is some tension around what to do when reconciling a service whose restart policy is "none". On one hand, the restart policy says not to restart a failed task. On the other hand, the declarative state of the service says a certain number of replics of the service should run, or the task should run on all possible nodes, and a change to the spec might resolve the condition that caused the failure.

Currently, the replicated orchestrator chooses to favor the declarative state when reconciling a service. This means that on a leader change or change to the service, we effectively ignore the restart policy.

The global orchestrator, on the other hand, refuses to touch the task on any node where it has failed, if the restart policy says to do so. This can lead to very unintuitive behavior, for example setting up a service that fails because its command line is wrong, then updating the service to fix the command line, and seeing nothing happen.

I believe the replicated orchestrator's behavior is a better choice, and less confusing to users. It's very bad for a service to get into a state where it can't be updated because it has already failed on all nodes and therefore won't be touched on those nodes.

This changes the global orchestrator to behave like the replicated orchestrator, and ignore the restart policy on startup, and when the service is updated.

Note that this means that on leader changes, failed global services will be restarted, which might be contrary to expectations. But it also means that a "service update" will be able to fix a broken global service regardless of the restart policy, which is very important.

cc @aluzzardi 
cc @dongluochen (in case you're still around)

stevvooe commented:
LGTM

This might interfere with some uses of restart-none as a batch mode primitive, but I am not sure we should support that hack in lieu of a real batch solution.

thaJeztah commented:
@stevvooe agreed; I was referring to the same in my comment

cyli commented:
Would it be possible to constrain restarting services on leader change if we check the task spec version (if it exists?)  So if the task is failed, but the service spec version hasn't changed, then even on leader change that particular task doesn't get restarted?

Other than that, the behavior and code seem good to me, with the caveat that I'm not super familiar with this code.  Perhaps it would be good to add a test case though?  There don't seem to be any tests for `RestartOnNone` policies.

aaronlehmann commented:
#932 affects both orchestrators. This PR makes the global orchestrator consistent with the replicated orchestrator, and whatever is done to fix #932 will have to change both in a consistent way. #2290 is a first attempt at this, which affects both orchestrators through the changes it makes to the restart supervisor. It also depends on the changes here for correctness.

> Would it be possible to constrain restarting services on leader change if we check the task spec version (if it exists?) So if the task is failed, but the service spec version hasn't changed, then even on leader change that particular task doesn't get restarted?

Something like this is possible, but it depends on the failed task still existing. We've tried to avoid depending on that, because sometimes users use `--task-history-limit 0` and it would be unexpected for that to influence orchestration behavior.

The solution in #2290 seems promising to me, but if it doesn't work out for some reason, this might be something to consider as an alternative.

cyli commented:
@aaronlehmann Ah ok.  Thanks for explaining.

aaronlehmann commented:
I've added some test coverage, fixed the tests not to leak the orchestrator, and changed the global orchestrator to ignore task deletions (which would wrongly trigger a reconciliation).

cyli commented on manager/orchestrator/global/global_test.go:
```diff
@@ -221,28 +258,72 @@ func TestDeleteService(t *testing.T) {
+	testutils.Expect(t, watch, state.EventCommit{})
+	testutils.Expect(t, watch, api.EventUpdateTask{}) // ready->running
+	testutils.Expect(t, watch, state.EventCommit{})
+
```
Would it make sense to also delete the task here, and assert that it doesn't get re-created, since that behavior has also changed?

aaronlehmann replied:
Added back `TestRemoveTask`, now expecting different behavior.

cyli commented:
One minor test nit, otherwise LGTM!

aluzzardi commented:
LGTM
